### PR TITLE
Azure VMware Solution 2023-09-01 in TypeSpec

### PR DESCRIPTION
**Pull request by cataggar**
This adds the 2023-09-01 API version and moves our spec to TypeSpec. It was ARM approved in https://github.com/Azure/azure-rest-api-specs-pr/pull/11977 and then ARM approved and breaking change approved in https://github.com/Azure/azure-rest-api-specs-pr/pull/15631. Future API versions will be easier to author and review. The changes in this API version include:

- addition of `IscsiPath` `ProxyResource` with `list`, `get`, `createOrUpdate`, and `delete` operations
- addition of `elasticSanVolume?` on `DatastoreProperties`
- addition of `nsxtManagerIp?`, `vcenterIp?`, and `hcxCloudMangerIp?` on `Endpoints`
- addition of `audience?` on `ScriptCmdletProperties`

# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

> [!NOTE]
> As of January 2024 there is no PR assignee. This is expected. See https://aka.ms/azsdk/pr-arm-review.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

diagram

<details>
<summary> Click here to see the details of Step 1 </summary>

### Breaking changes review (Diagram Step 1)

If the automation determines you have breaking changes, i.e. Step 1 from the diagram applies to you,
  you must follow the [breaking changes process].  
  **IMPORTANT** This applies even if:
- The tool fails while it shouldn't, e.g. due to runtime exception, or incorrect detection of breaking changes.
- You believe there is no need for you to request breaking change approval, for any reason. 
    Such claims must be reviewed, and the process is the same.

</details>

<details> 
<summary> Click here to see the details of Step 2 </summary>

### ARM API changes review (Diagram Step 2)

- For help with Step 2 / "ARM review", see https://aka.ms/azsdk/pr-arm-review.

</details>

<details>
<summary> Click here to see the diagram footnotes</summary>

### Diagram footnotes

[1] See [ARM review queue] (for **PR merge** queues, see [2]).  
[2] [public repo merge queue], [private repo merge queue] (for **ARM review** queue, [1])  
The ARM reviewer on-call engineer visits the merge queue twice a day, so the approximate ETA for merges is 12 - 24 hours.

</details>

## Purpose of this PR

What's the purpose of this PR? Check the specific option that applies. This is **mandatory**!

  - [ ] New resource provider.
  - [x] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been generated using [OpenAPI Hub](https://aka.ms/openapihub)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with PR workflow diagram Step 2 (ARM review), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

[ARM review queue]: https://aka.ms/azsdk/pr-arm-review#about-the-arm-review-queue
[public repo merge queue]: https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+draft%3Afalse+sort%3Acreated-asc
[private repo merge queue]: https://github.com/Azure/azure-rest-api-specs-pr/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+-label%3AApproved-OkToMerge+draft%3Afalse+sort%3Acreated-asc
[breaking changes process]: https://eng.ms/docs/cloud-ai-platform/azure-core/azure-core-pm-and-design/trusted-platform-pm-karimb/service-lifecycle-and-actions-team/service-lifecycle-actions-team/apex/media/launchingproductbreakingchanges#breaking-change-process-1

### CODE REVIEW

File: specification/vmware/Microsoft.AVS/models.tsp
Comment (cataggar) [inline specification/vmware/Microsoft.AVS/models.tsp:2322]: The addition of `@singleton` occurred after the most recent ARM approval.

**Comment by cataggar** [comment]
The LintDiff EvenSegmentedPathForPutOperation errors look like false positives, so I created https://github.com/Azure/azure-sdk-tools/issues/7801 .

**Comment by cataggar** [comment]
The LintDiff PatchBodyParametersSchema errors also look like false positives, so I created https://github.com/Azure/azure-sdk-tools/issues/7802.

**Comment by cataggar** [comment]
For the Breaking Change(Cross-Version) error, I believe it is also a false positive. 
https://github.com/Azure/azure-sdk-tools/issues/7803

This was previously discussed in a breaking change review and approved.

**Comment by cataggar** [comment]
Currently in position 10 or 31 for the [ARM review queue](https://github.com/search?q=org%3AAzure+is%3Apr+is%3Aopen+label%3AWaitForARMFeedback+-label%3AIDCDevDiv++draft%3Afalse+sort%3Acreated-asc+&type=pullrequests). The estimate is about 10 reviews a day, so it may be reviewed today.

**Comment by cataggar** [comment]
Currently in position 9 of 42 in the ARM review queue. Moved up only 1 position since yesterday.

File: specification/vmware/Microsoft.AVS/models.tsp
Comment (TimLovellSmith) [reply specification/vmware/Microsoft.AVS/models.tsp:2085]: And that is okay to add because it always was, really, a singleton, is it?

File: specification/vmware/Microsoft.AVS/models.tsp
Comment (TimLovellSmith) [inline specification/vmware/Microsoft.AVS/models.tsp:25]: >quote [](http://example.com/codeflow?start=6&length=5)

Quota?

Comment (cataggar) [reply specification/vmware/Microsoft.AVS/models.tsp]: Yes, fixed in d0af234a87db6c8b2ffb83da6ce4c72299a3a2ce.

File: specification/vmware/Microsoft.AVS/models.tsp
Comment (cataggar) [reply specification/vmware/Microsoft.AVS/models.tsp:2085]: Yes, I believe it should be. I just missed it in the port to TypeSpec. It was introduced in Swagger format in https://github.com/Azure/azure-rest-api-specs-pr/pull/13757 . @balram2697, can you please confirm that this should be a `@singleton`?

File: specification/vmware/Microsoft.AVS/examples/2023-09-01/WorkloadNetworks_Get.json
Comment (TimLovellSmith) [inline specification/vmware/Microsoft.AVS/examples/2023-09-01/WorkloadNetworks_Get.json:9]: >workloadNetworkName": "default" [](http://example.com/codeflow?start=5&length=31)

just wondering, is it meant to still be needed? Or is it a singleton?

File: specification/vmware/Microsoft.AVS/models.tsp
Comment (TimLovellSmith) [reply specification/vmware/Microsoft.AVS/models.tsp:2100]: not really important though

File: specification/vmware/Microsoft.AVS/examples/2023-09-01/WorkloadNetworks_Get.json
Comment (cataggar) [reply specification/vmware/Microsoft.AVS/examples/2023-09-01/WorkloadNetworks_Get.json:9]: It is a singleton. I'll remove those entries.

**Comment by TimLovellSmith** [comment]
>             "type": "object"

is this completely free form?

---
Refers to: specification/vmware/resource-manager/Microsoft.AVS/stable/2023-09-01/vmware.json:9382 in 1c7a144. [](commit_id = 1c7a144fb6df56ea4c24bd865de25fdc524908b6, deletion_comment = False)

File: specification/vmware/resource-manager/readme.md
Comment (TimLovellSmith) [inline specification/vmware/resource-manager/readme.md:278]: >records [](http://example.com/codeflow?start=41&length=7)

what records are we talking about here?

File: specification/vmware/resource-manager/readme.md
Comment (TimLovellSmith) [inline specification/vmware/resource-manager/readme.md:281]: >XmsPageableForListCalls [](http://example.com/codeflow?start=10&length=23)

I seem to remember one of the reasons to include it is in case ARM injects this into the list response somehow.

File: specification/vmware/resource-manager/readme.md
Comment (cataggar) [reply specification/vmware/resource-manager/readme.md:278]: Probably related to https://github.com/Azure/azure-sdk-tools/issues/7803 .

File: specification/vmware/resource-manager/readme.md
Comment (TimLovellSmith) [inline specification/vmware/resource-manager/readme.md:292]: >ArmResourceCreateOrUpdateAsync [](http://example.com/codeflow?start=24&length=30)

it seems like all the properties of IscisiPath are readonly, am I understanding it right? Then how does it have a create/update action?

File: specification/vmware/Microsoft.AVS/routes.tsp
Comment (TimLovellSmith) [inline specification/vmware/Microsoft.AVS/routes.tsp:893]: >IscsiPath [](http://example.com/codeflow?start=51&length=9)

it doesn't seem to have any writable properties? is it creatable anyway?

File: specification/vmware/resource-manager/readme.md
Comment (cataggar) [reply specification/vmware/resource-manager/readme.md:292]:
```
  @doc("CIDR Block for iSCSI path.")
  @visibility("create", "read")
  networkBlock: string;
```

File: specification/vmware/Microsoft.AVS/routes.tsp
Comment (cataggar) [reply specification/vmware/Microsoft.AVS/routes.tsp:893]: `networkBlock` has `create`

**Comment by cataggar** [comment]
> > ```
> >         "type": "object"
> > ```
> 
> is this completely free form?
> 
> Refers to: specification/vmware/resource-manager/Microsoft.AVS/stable/2023-09-01/vmware.json:9382 in [1c7a144](https://github.com/Azure/azure-rest-api-specs/commit/1c7a144fb6df56ea4c24bd865de25fdc524908b6). [](commit_id = [1c7a144](https://github.com/Azure/azure-rest-api-specs/commit/1c7a144fb6df56ea4c24bd865de25fdc524908b6), deletion_comment = False)

I think this comment is referring to:
``` json
        "namedOutputs": {
          "type": "object",
          "description": "User-defined dictionary.",
          "additionalProperties": {
            "type": "object"
          }
        },
```
This matches the `2023-03-01` API spec. Not a direct answer.

File: specification/vmware/Microsoft.AVS/models.tsp
Comment (prtyag) [reply specification/vmware/Microsoft.AVS/models.tsp:2085]: Yes IscsiPath is the singleton.

**Review by prtyag — APPROVED** [review]
Changes wrt  IScsiPath and Datastore LGTM.

File: specification/vmware/resource-manager/readme.md
Comment (cataggar) [reply specification/vmware/resource-manager/readme.md:292]: Will close. `networkBlock` is set during a create.

File: specification/vmware/Microsoft.AVS/examples/2023-09-01/WorkloadNetworks_Get.json
Comment (cataggar) [reply specification/vmware/Microsoft.AVS/examples/2023-09-01/WorkloadNetworks_Get.json]: removed in 2213585eed54328a19f28437c756ba3717600095

File: specification/vmware/Microsoft.AVS/models.tsp
Comment (cataggar) [reply specification/vmware/Microsoft.AVS/models.tsp:2100]: @prtyag, what do you think?

Comment (cataggar) [reply specification/vmware/Microsoft.AVS/models.tsp:2100]: What about using the pattern from `podCidr` or `serviceCidr` from ContainerService?

``` json
"pattern": "^([0-9]{1,3}\\.){3}[0-9]{1,3}(\\/([0-9]|[1-2][0-9]|3[0-2]))?$",
```

https://github.com/Azure/azure-rest-api-specs/blob/84849e5293de3a8e01cbdde13b2d7086d3fb34d6/specification/containerservice/resource-manager/Microsoft.ContainerService/aks/stable/2024-01-01/managedClusters.json#L3746-L3757

**Comment by cataggar** [comment]
Received `Approved-SdkBreakingChange-Python` yesterday. Waiting on [CI-BreakingChange-Go](https://github.com/Azure/azure-rest-api-specs/labels/CI-BreakingChange-Go) & [CI-BreakingChange-JavaScript](https://github.com/Azure/azure-rest-api-specs/labels/CI-BreakingChange-JavaScript).

**Comment by kazrael2119** [comment]
don't know why breaking change cross-version check got failure, JS sdk breaking may caused in api level, so I can't add JS approval

**Comment by cataggar** [comment]
> don't know why breaking change cross-version check got failure, 

That is a false positive. I created https://github.com/Azure/azure-sdk-tools/issues/7803.

image

> JS sdk breaking may caused in api level, so I can't add JS approval

For the JS SDK, you can use the `Generated ApiView` above to see that it is not breaking.

image

@kazrael2119, please approve or give more details on why you are unable to.

**Comment by cataggar** [comment]
https://apiview.dev/Assemblies/Review/993c23ef90ad48f1a7e3dba3509092d5 is from the [2023-03-01 review.](https://github.com/Azure/azure-rest-api-specs/pull/24490). 

`tags` has the same API in TypeScript in the previous API version.

image

**Comment by cataggar** [comment]
Still waiting for a review of the Go SDK breaking changes and an approval for the JavaScript SDK breaking changes. 2 business days have passed since the March 6th ARM approval.

File: specification/vmware/Microsoft.AVS/models.tsp
Comment (prtyag) [reply specification/vmware/Microsoft.AVS/models.tsp:2100]: pattern is not important here.

**Comment by cataggar** [comment]
The bump to TypeSpec v0.54.0 resulted in additional warnings. I suppressed them as it is already ARM approved.

File: specification/vmware/resource-manager/readme.md
Comment (cataggar) [reply specification/vmware/resource-manager/readme.md:214]: This suppression has been removed.

File: specification/vmware/resource-manager/readme.md
Comment (cataggar) [reply specification/vmware/resource-manager/readme.md]: This suppression has been removed.

File: specification/vmware/Microsoft.AVS/routes.tsp
Comment (cataggar) [inline specification/vmware/Microsoft.AVS/routes.tsp:782]: May be `final-state-via` should be `location`.

> location - poll until terminal state, if the initial response had a Location header, a final GET will be done. Default behavior for POST operation.
https://github.com/Azure/autorest/tree/main/docs/extensions#x-ms-long-running-operation-options

Comment (cataggar) [reply specification/vmware/Microsoft.AVS/routes.tsp]: I looks like the default behavior for POST is location. Out of caution, I'm simply removing unsetting it, same as previous API version.

**Comment by cataggar** [comment]
@TimLovellSmith, can I please get a suppression approval?

**Comment by cataggar** [comment]
I added back the `WaitForARMFeedback` tag. I need both `ARMSignedOff` & `Approved-Suppression` from the ARM review like [before](https://github.com/Azure/azure-rest-api-specs-pr/pull/15631#event-11284610799).

**Comment by cataggar** [comment]
In position 6 of 20 in the [ARM review queue](https://github.com/search?q=org%3AAzure+is%3Apr+is%3Aopen+label%3AWaitForARMFeedback+-label%3AIDCDevDiv++draft%3Afalse+sort%3Acreated-asc+&type=pullrequests&p=1). It is sorted by `created-asc`.

File: specification/vmware/resource-manager/readme.md
Comment (raosuhas) [inline specification/vmware/resource-manager/readme.md:228]: path please

File: specification/vmware/resource-manager/readme.md
Comment (raosuhas) [inline specification/vmware/resource-manager/readme.md:204]: could you please add the path ?

**Comment by raosuhas** [comment]
Changes look fine could you please add specific suppressions rahter than for the whole file for the suppression review ?

**Comment by cataggar** [comment]
All suppressions have paths. LintDiff is passing. Please grant suppression approval. I would like this PR merged.

**Review by raosuhas — APPROVED** [review]

**Pull request merged**
